Add next-gen Conv img migration doc (#5590)
circleci/circleci-docs: jekyll/_cci2/next-gen-migration-guide.md

---
layout: classic-docs
title: "Migrating to next-gen Convenience Images"
short-title: "Migrating to next-gen Convenience Images"
description: "Migrating to next-gen Convenience Images"
order: 30
version:
- Cloud
- Server v2.x
---

* TOC
{:toc}

## Overview

> Last year (2020) CircleCI introduced the next generation (next-gen) of Convenience Images. These new images are designed to replace the legacy Convenience Images that came out during the announcement of CircleCI 2.0, about 4 years ago. The next-gen CircleCI Convenience Images are designed from the ground up for a Continuous Integration environment. They’re designed to be faster, more efficient, and more importantly, more reliable. You can learn more about all of the features on our blog post. As we begin to deprecate the legacy images, here’s information on the migration process.

Reviewer: This is reads a bit more like a blog post - should it be a blog post that we can link to from a doc. Specifically, I think including historical data (that is is not evergreen, ie, we will need to come back and change `4 years ago` to `five years ago` in a year etc.).

Reviewer 2: ```suggestion
Last year (2020) CircleCI introduced the next generation (next-gen) of Convenience Images. These new images are designed to replace the legacy Convenience Images that came out during the announcement of CircleCI 2.0, about 4 years ago. The next-gen CircleCI Convenience Images are designed from the ground up for a Continuous Integration environment. They’re designed to be faster, more efficient, and more importantly, more reliable. You can learn more about all of the features on our blog post. As we begin to deprecate the legacy images, here is information on the migration process.
```

> Last year (2020) CircleCI introduced the next generation (next-gen) of Convenience Images. These new images are designed to replace the legacy Convenience Images that came out during the announcement of CircleCI 2.0, about 4 years ago. The next-gen CircleCI Convenience Images are designed from the ground up for a Continuous Integration environment. They’re designed to be faster, more efficient, and more importantly, more reliable. You can learn more about all of the features on our blog post. As we begin to deprecate the legacy images, here is information on the migration process.

Reviewer 3: ```suggestion
In 2020 CircleCI introduced the next generation (next-gen) of convenience images. These new images are designed to replace the legacy convenience images that came out during the announcement of CircleCI 2.0, around 2016. The next-gen CircleCI convenience images are designed from the ground up for a CI/CD environment. These images are designed to be faster, more efficient, and, most importantly, more reliable. You can learn more about these features in our blog post. As we begin to deprecate the legacy images, this page outlines information on the migration process.
```

Reviewer 3: @felicianotech can we link to the blog post here?

Author: There two links/blog posts here. I added both back.

> Moving from a legacy to next-gen image is easy and straightforward. All legacy images have a Docker namespace of “circleci” while next-gen images have a Docker namespace of “cimg”. For example, migrating from the legacy Ruby or Python image to the respective next-gen:

Reviewer: Avoid use of "easy" or "straightforward".

Author: So what should I put? Should I remove the sentence altogether?

Reviewer 3: I've made a suggested change for this in my review below

> Moving from a legacy to next-gen image is easy and straightforward. All legacy images have a Docker namespace of “circleci” while next-gen images have a Docker namespace of “cimg”. For example, migrating from the legacy Ruby or Python image to the respective next-gen:

Reviewer 3: ```suggestion
Moving from a legacy to next-gen image requires a change to the namespace. All legacy images have a Docker namespace of `circleci`, while next-gen images have a Docker namespace of `cimg`. For example, migrating from the legacy Ruby or Python image to the respective next-gen image can be done as follows:
```

circleci/ruby:2.3.0 -> cimg/ruby:2.3.0
circleci/python:3.8.4 -> cimg/python:3.8.4

## Changes

> ### Deprecated images

Reviewer: each MD header needs a custom kramdow id-slug so that our docs can be translated. So for example, this would be named:

```
### Depgrecated images
{: #deprecated-images }
```

Author: I can definitely do it, but I am curious. Jekyll already does this. The translating part can't just use that?

Reviewer 3: @felicianotech yeah I've wondered this too 🤷‍♀️

Reviewer: It's a bit of a manual process, but for I18N purposes, source docs (ie, english) have specified manual id's so that target languages (ie, just japanese for now) don't auto-generate the ids; if we don't do this, I18n target languages aren't able to inter-link to anchor headings as the translators weren't translating html link tags. 

An example:

![image](https://user-images.githubusercontent.com/12987958/130496207-bbe15c83-5cf3-4278-93f4-7545e0718aec.png)

Author: Gotcha. I'll add these in in my list commit.

Author: Done

> There are a few images that will be completely deprecated, without a replacement, as well as one that has a name change. The buildpack-deps, JRuby, and DynamoDB images will be fully deprecated with no next-gen equivalents. If you are using the buildpack-deps image, the suggestion is to use the new CircleCI Base image, cimg/base. For the other two images, you can install the software yourself in the base image or use a 3rd-party image instead. The Go image name has been changed from golang to go. All of the Legacy to Next-Gen image changes are captured below in this table:

Reviewer 2: ```suggestion
Several changes are being made to existing images. The following images will be deprecated without next-gen equivalents:

* `buildpack-deps`
* `JRuby`
* `DynamoDB`

If you are using the `buildpack-deps` image, the suggestion is to use the new CircleCI Base image, `cimg/base`. For the other two images, you can install the software yourself in the base image or use a 3rd-party image instead.

Also, the following image will be renamed:

* The Go image will be changed from `golang` to `go`

All Legacy to Next-Gen image changes are captured below in this table:

```

| Legacy Image | Next-Gen Image |
| --- | --- |
| circleci/buildpack-deps | cimg/base |
| circleci/jruby | No suggested path |
| circleci/dynamodb | No suggested path |
| circleci/golang | cimg/go |

> 

Reviewer 3: ```suggestion
{: class="table table-striped"}
```

### Browser Testing

> With legacy images, there were 4 different variant tags you could use to get browser testing for a particular image. For example, if you were doing browser testing with the Python v3.7.0 image, you might have used Docker image: circleci/python:3.7.0-browsers. These 4 tags have now been consolidated into a single tag designed to be used together with the CircleCI Browser Tools orb.

Reviewer: Maybe we could link to the browser tools orb?

Author: I did but it was lost. Good catch. Adding it back in.

Legacy Variant Tags

Next-gen Variant Tag
-browsers
-browsers + Browser Tools Orb
-browsers-legacy
-node-browsers

> -node-browsers-legacy

Reviewer: Not sure if I'm misunderstanding, but are these mixed up? Should there be something under the "Legacy Variant Tags" section? Should these be list-items? If so, should -browsers be `-browsers` with ?

Reviewer 3: Yes I am wondering the same?

Author: Yeah this is messed up. It was a table that lost its formatting when the guide was converted over. I am fixing it now and looking for other places where these conversion problems might have occured.

> The new, single browsers variant tag includes Node.js, common utilities to use in browser testing such as Selenium, but not the actual browsers. Please note the browsers are no longer pre-installed. Instead, browsers such as Google Chrome and Firefox as well as their drivers Chromedriver and Gecko are installed via the browsers-tools orb. This provides the flexibility to mix and match the browser versions you need in a build rather than using strictly what CircleCI provides. You can find examples of how to use this orb here.

Reviewer: > You can find examples of how to use this orb here.

Should `here` have a link to somewhere?

Reviewer 2: ```suggestion
The new, single browsers variant tag includes Node.js, common utilities to use in browser testing such as Selenium, but not the actual browsers. Please note the browsers are no longer pre-installed. Instead, browsers such as Google Chrome and Firefox, as well as their drivers Chromedriver and Gecko, are installed via the `browsers-tools` orb. This provides the flexibility to mix and match the browser versions you need in a build rather than using strictly what CircleCI provides. You can find examples of how to use this orb here.
```

Using Ubuntu as the Single Base OS

> Legacy images had variant tags for several different base operating systems (OS). Some images were versions of Debian and Ubuntu while other images provided several different bases. This is no longer the case. All CircleCI next-gen images are based on the latest LTS release of Ubuntu.

Reviewer 2: ```suggestion
Legacy images had variant tags for several different base operating systems. Some images were versions of Debian and Ubuntu, while other images provided several different bases. This is no longer the case. All CircleCI next-gen images are based on the latest LTS release of Ubuntu.
```

With the base image, at least two LTS releases and non-EOL’d standard releases will be supported.

## Troubleshooting

> When migrating to a next-gen image, there might be some software issues. Common issues include a library you were using now has a different version or an apt package is no longer pre-installed. In this scenario simply install that package using:

Reviewer 3: ```suggestion
When migrating to a next-gen image, there might be some software issues. Common issues include:
* A library you were using now has a different version. 
* An apt package is no longer pre-installed. In this scenario simply install that package using:
```

> sudo apt-get update && sudo apt-get install -y <the-package>

Reviewer: Nit: This should be in a code block with shell highlighting.

Author: done

> Every image has its own GitHub repository. You can find them here. These repositories are where you can learn more about an image, what makes up the image, open GitHub Issues, and contribute PRs. If you’re having an issue with an image, especially if it’s a migration issue, you can open up a GitHub Issue and ask questions. You can always open up a Support ticket or reach out on CircleCI Discuss. 

Reviewer 2: I think we may need a link to the image GH repository where "here" is stated. Also, should GitHub Issues be capitalized? Just a minor nit on that one. Finally, should we link to how to open a support ticket or post in Discuss?

Reviewer 2: ```suggestion
Every image has its own GitHub repository. You can find them here. These repositories are where you can learn more about an image, what makes up the image, open GitHub Issues, and contribute PRs. If you are having an issue with an image, especially if it’s a migration issue, you may open up a GitHub Issue and ask questions. You also may open a Support ticket or reach out on CircleCI Discuss. 
```

Reviewer 3: GitHub do not capitalise "issues" in their docs so we shouldn't.

Author: This is something I do to differentiat when I as talking about an "issue" with software or something versus a GitHub "issue", the feature on their website. I changed it to lowercase.

---
layout: classic-docs
title: "Migrating to next-gen Convenience Images"
short-title: "Migrating to next-gen Convenience Images"

> description: "Migrating to next-gen Convenience Images"

Reviewer 3: ```suggestion
description: "A guide for migrating from using legacy images to next-gen images."
```

order: 30
version:
- Cloud
- Server v2.x
---

* TOC
{:toc}

## Overview

> Last year (2020) CircleCI introduced the next generation (next-gen) of Convenience Images. These new images are designed to replace the legacy Convenience Images that came out during the [announcement of CircleCI 2.0](https://circleci.com/blog/say-hello-to-circleci-2-0/), about 4 years ago. The next-gen CircleCI Convenience Images are designed from the ground up for a Continuous Integration environment. They’re designed to be faster, more efficient, and more importantly, more reliable. You can learn more about all of the features [on our blog post](https://circleci.com/blog/announcing-our-next-generation-convenience-images-smaller-faster-more-deterministic/). As we begin to deprecate the legacy images, here is information on the migration process.

Reviewer 3: ```suggestion
In 2020 CircleCI introduced the next generation (next-gen) of convenience images. These new images are designed to replace the legacy convenience images that were released during the [announcement of CircleCI 2.0](https://circleci.com/blog/say-hello-to-circleci-2-0/). The next-gen CircleCI convenience images are designed from the ground up for a CI/CD environment. They are designed to be faster, more efficient, and most importantly, more reliable. You can learn more about all of the features [on our blog post](https://circleci.com/blog/announcing-our-next-generation-convenience-images-smaller-faster-more-deterministic/). As we begin to deprecate the legacy images, this document provides information on the migration process.
```

Also, the following image will be renamed:

* The Go image will be changed from `golang` to `go`

All Legacy to Next-Gen image changes are captured below in this table:

| Legacy Image | Next-Gen Image |
| --- | --- |
| circleci/buildpack-deps | cimg/base |
| circleci/jruby | No suggested path |
| circleci/dynamodb | No suggested path |
| circleci/golang | cimg/go |
{: class="table table-striped"}

### Browser Testing

With legacy images, there were 4 different variant tags you could use to get browser testing for a particular image. For example, if you were doing browser testing with the Python v3.7.0 image, you might have used Docker image: circleci/python:3.7.0-browsers. These 4 tags have now been consolidated into a single tag designed to be used together with the [CircleCI Browser Tools orb](https://circleci.com/developer/orbs/orb/circleci/browser-tools).

<table>
<tr><th>Legacy Variant Tags</th><th>Next-gen Variant Tags</th></tr>
<tr><td>-browsers</td><td rowspan=4>-browsers + Browser Tools Orb</td></tr>
<tr><td>-browsers-legacy</td></tr>
<tr><td>-node-browsers</td></tr>
<tr><td>-node-browsers-legacy</td></tr>

> </table>

Reviewer 3: This table needs to be formatted like the one in the section above

Author: It can't. The table above is a basic table so it can be Markdown formatted. This table has a rowspan which cannot be done in GitHub Flavored Markdown, which is what we use in Docs.

Author: I got two approvals so I am going to go ahead and merge. Please let me know if you'd like to discuss this further.

Using Ubuntu as the Single Base OS

Legacy images had variant tags for several different base operating systems (OS). Some images were versions of Debian and Ubuntu while other images provided several different bases. This is no longer the case. All CircleCI next-gen images are based on the latest LTS release of Ubuntu.

With the base image, at least two LTS releases and non-EOL’d standard releases will be supported.

## Troubleshooting

When migrating to a next-gen image, there might be some software issues. Common issues include:
* A library you were using now has a different version. 
* An apt package is no longer pre-installed. In this scenario simply install that package using:

```bash
sudo apt-get update && sudo apt-get install -y <the-package>
```

> Every image has its own GitHub repository. You can find them [here](https://github.com/CircleCI-Public?q=cimg-&type=&language=&sort=). These repositories are where you can learn more about an image, what makes up the image, open GitHub Issues, and contribute PRs. If you are having an issue with an image, especially if it’s a migration issue, you may open up a GitHub Issue and ask questions. You also may open a Support ticket or reach out on CircleCI Discuss. 

Reviewer 3: ```suggestion
Each image has its own GitHub repository. You can find them [here](https://github.com/CircleCI-Public?q=cimg-&type=&language=&sort=). These repositories are where you can learn more about an image, what makes up the image, open GitHub issues, and contribute PRs. If you are having an issue with an image, especially if it is a migration issue, you can open up a GitHub issue and ask questions. You can also open a support ticket or reach out on CircleCI Discuss. 
```

Reviewer 3: Can we add links to support and discuss here?
